### ISSUE TITLE
inDatabase Method crash

### ISSUE THREAD

**Issue by AnYuan** [ccgus/fmdb#404]
```objc
#if defined(DEBUG) && DEBUG
            NSSet *openSetCopy = FMDBReturnAutoreleased([[db valueForKey:@"_openResultSets"] copy]);
            for (NSValue *rsInWrappedInATastyValueMeal in openSetCopy) {
                FMResultSet *rs = (FMResultSet *)[rsInWrappedInATastyValueMeal pointerValue];
                NSLog(@"query: '%@'", [rs query]);
            }
#endif
```

My app crash at
 
```objc 
NSLog(@"query: '%@'", [rs query]);
```

debugger tells me that I was sending query method to a NSString object which does not recognize the method.

Anyone knows why this happens?

Thanks a lot.

**Comment by ccgus** [comment]
Can you provide sample code reproducing the crash, along with a stack trace?

**Comment by Amitmundra** [comment]
Can you please tell me, why do we need below line of code, we are already calling releasing & FMDBReturnAutoreleased.
Because below code is crashing the app.
 for (NSValue *rsInWrappedInATastyValueMeal in openSetCopy) {
                FMResultSet *rs = (FMResultSet *)[rsInWrappedInATastyValueMeal pointerValue];
                NSLog(@"query: '%@'", [rs query]);
            }

**Comment by robertmryan** [comment]
I assume this is related to issue https://github.com/ccgus/fmdb/issues/131, where you posted this image:

<img src="https://cloud.githubusercontent.com/assets/11942529/19432039/68a64296-9478-11e6-9b3b-d7eb9c395962.png">

As I asked there, are you doing ARC or manual reference counting? If manual reference counting, did you run it through the static analyzer ("Analyze" on Xcode's "Product" menu)? You should have a clean bill of health from that before we go further.

I'd also be inclined to turn on zombies, as this feels like a dangling pointer problem.

I focus on these diagnostic procedures because you appear to be dwelling on this code in the `inDatabase` routine that has been used by probably lots of developers since it was introduced in 2013. I suspect your problem is just a symptom other, deeper, problem in how you're using FMDB, not a problem in FMDB, itself.

**Comment by Amitmundra** [comment]
Thanks for reply robertmryan,
we are using ARC.
let me tell you one thing when i comment these 3 line of code then crashing goes away, but we are calling FMDBReturnAutoreleased method but not doing any thing with returned  NSSet *openSetCopy. Please please help i am stuck here, i don't know why is this error.

for (NSValue *rsInWrappedInATastyValueMeal in openSetCopy) {
    FMResultSet *rs = (FMResultSet *)[rsInWrappedInATastyValueMeal pointerValue];
    NSLog(@"query: '%@'", [rs query]);
}

Can you please tell me why do we need above 3 line of code?

Thanks.

**Comment by robertmryan** [comment]
Obviously, if you do a release build, this code won't be executed. Just edit your scheme accordingly. 

But this is addressing the symptom of a problem, not the problem itself. Whatever is causing this problem is likely to cause problems even if you do a release build (which excludes this code).

So I would suggest two things:

1. But I wonder how you're possibly getting to that portion of the code at all. It means that your code is hanging on to a `FMResultSet` outside of the scope of the `inDatabase`. Don't do that. The `FMResultSet` variables should always be local variables inside the blocks.

2. I don't know how your code is crashing there, because when I force my code to call this (by incorrectly keeping external strong reference to `FMResultSet`), it works fine. Again, I'd suggest temporarily turning on zombies to see if this problem is a result of a dangling pointer. I'd try turning on the address sanitizer, or if that fails to identify anything, some of the "malloc" diagnostic settings, too.

And, again, if the problem persists, we need a [reproducible example of the problem](http://stackoverflow.com/help/mcve). Otherwise we can't help you. Generally, if we can't reproduce the problem, we can't solve the problem.

**Comment by robertmryan** [comment]
By the way, I don't think I answered your question:

> Can you please tell me why do we need above 3 line of code?

The purpose of those lines of code are to report diagnostic information (in debug builds only) about any result set objects that haven't been deallocated by the time the `inDatabase` call is done. So, either you have some external reference to a `FMResultSet` object that you created from within the block, or your doing some asynchronous calls from within the block that is using the result set.

**Comment by Amitmundra** [comment]
Thanks for reply Mr robertmryan,
I run the app without debug mode now crashing stop. one more question
i have 3 different thread 2 run in background and 1 run in foreground.
when all 3 thread does some database operation then it slow the UI, Let say if 2 background thread are doing some db operation and same time UI is demanding for data then it hang the UI. Background threads are using inTransection for db operation, can you help me how can i avoid UI hang in these cases?

Thanks.

**Comment by robertmryan** [comment]
The fundamental issue is that you have three threads contending for a slow,shared, synchronized resource. 

- So, I'd first review the background threads to see how to minimize the amount of time they take. Make sure you don't have any synchronous tasks inside the `inTransaction` blocks. I'd start with with benchmarking all of them and see if there are any that are taking a disproportionate amount of time and focus my attention on seeing if I can get in and out of the `inTransaction` blocks as quickly as possible (without radically using too many `inTransaction` blocks, thereby losing the benefit of transactions). 

Usually this is enough to solve the problem. But, when it's not (e.g. the background thread needs to update hundreds of thousands of records and there's no way around that), I then consider:

- In those cases where there are a ton of background updates (i.e. even after optimizing the background threads threads, they're still tying up the `FMDatabaseQueue` for longer than I'd like), I then (a) break up those background operations into smaller components that don't make, for example, more than 500msec per block; and (b) set up some mechanism that prioritizes foreground requests over background requests (e.g. wrap the `FMDatabaseQueue` in a serial operation queue and give the foreground operations a higher priority than the background ones. 

 This gets a little complicated, so I only do this when _absolutely_ necessary.

**Comment by robertmryan** [comment]
BTW, you said: 

> I run the app without debug mode now crashing stop.

Not to belabor the issue, but while that's good, you're still just addressing the symptom rather than the root problem. The question in my mind is why you have open result sets when the `inTransaction` is done. I'm gathering that you don't want to tackle that right now (and, I get it, sometimes we have to set aside this sort of design issue to take care of more pressing issues), but I'd really want to figure out why those result sets are not getting closed, and fix that root problem. 

I'd also be concerned that the code was crashing when doing a debug build. This suggests a second, deeper memory management problem. So, I'd go back to a debug build, and then tracking down the problem through a combination of temporarily turning on zombies, the address sanitizer, and doing static analysis of the code. By simply turning off debug mode, you may simply be hiding some more fundamental memory problem that may manifest itself later in unpredictable fashion.

Again, I understand you might not want to tackle that right now, but don't be surprised if you start seeing other unpredictable problems rearing their ugly head in the future. Often, you'll save yourself time in the long run if you tackle the root problem while you have a nice, easily reproducible example of the issue. 

Just my two cents.
